### ISSUE TITLE
Test "test_batch_normalization" segfaults just after being called, in a wide range of scenarios

### ISSUE THREAD

**Issue by emaballarin** [intel/mkl-dnn#223]
Hello,

I just tried to compile MKL-DNN. I am somehow a longtime user of the library, and I usually succeeded in building and testing it with this same setup (the last time, however, was months ago).

Now, with a fresh reinstall of the OS, the build succeeds but one of the tests fails. I tried to compile with different compiler and libraries solutions... all with the same result.

I have tried to document thoroughly the error below, that happens both with GNU and Intel compilers, with or without MKL and/or IntelMP support.

I am obviously available for further testing/providing more info if needed.

Thanks for your work and support.

Regards.

-----------------------------------------------------------------------------

### Environment
* CPU manufacturer and model
```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  2
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               94
Model name:          Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
Stepping:            3
CPU MHz:             900.067
CPU max MHz:         3600.0000
CPU min MHz:         800.0000
BogoMIPS:            5426.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            8192K
NUMA node0 CPU(s):   0-7
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves ibpb ibrs stibp dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
```

* OS version
```
Linux emaballarin-precision-workstation 4.14.34-1-MANJARO #1 SMP PREEMPT Thu Apr 12 17:26:43 UTC 2018 x86_64 GNU/Linux
```
_Note:_ it's just the Arch Linux kernel with preemption support.
* Compiler version
_Tried both GCC and Intel C/C++, with versions as follows..._
```
$ gcc --version

gcc (GCC) 7.3.1 20180312
Copyright (C) 2017 Free Software Foundation, Inc.
```
```
$ icc --version

icc (ICC) 19.0.0.046 20180317
Copyright (C) 1985-2018 Intel Corporation.  All rights reserved.
```
* MKLROOT value and configuration
```
$ echo MKLROOT=$MKLROOT

MKLROOT=/opt/intel/compilers_and_libraries_2019.0.046/linux/mkl
```

_From ~/.bashrc..._
```
source /opt/intel/parallel_studio_xe_2019/bin/psxevars.sh
source /opt/intel/compilers_and_libraries/linux/mpi/intel64/bin/mpivars.sh
source /opt/intel/mkl/bin/mklvars.sh intel64
source /opt/intel/tbb/bin/tbbvars.sh intel64
export MKL_ROOT="$MKLROOT"
export MKLINC="/opt/intel/mkl/include"
export TBB_ROOT="$TBBROOT"
export MKLROOT="$MKL_ROOT"
export LD_LIBRARY_PATH="/opt/intel/compilers_and_libraries_2019.0.046/linux/compiler/lib/intel64_lin:$LD_LIBRARY_PATH"
export MKL_THREADING_LAYER=INTEL
export OMP_NUM_THREADS=4
export KMP_AFFINITY=verbose,granularity=fine,proclist=[0-4],explicit
```
_Note: however, the problem occurs exactly the same when compiling without Intel MKL and IntelMP support. I have manually checked that the OpenMP library used is the one wanted (I had in the past libiomp/libgomp incompatibilites, but it was on a completely different setup)._

* CMake version
```
cmake version 3.10.3
CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

* CMake output log
I will actually have a hard time finding the build log (I will provide it later today, if needed... sorry!).
However, the build went fine with both compilers, as expected. Please, see below (at the end) for maybe useful logfiles.

### Steps to reproduce
* Clone the mkl-dnn repository from git;
* Follow the readme instructions, **but without the MKL Small Library, building with (1) full MKL support or (2) no MKL support whatsoever**. The problem occurs with no difference.
* Compile with (a) Intel Parallel Studio XE compilers or (b) GNU Compilers Suite. The problem occurs with no difference.

### Actual behavior
* The tests are run;
* Test N° 21/36 "test_batch_normalization" fails with **Exception: segmentation fault**. All the other tests pass.

### Expected behavior
* The tests are run;
* All tests pass.

### Additional information:
* **The configuration file, sourced to setup the Intel Compilers toolchain**
```
# C/C++/Fortran
export CC=icc
export CXX=icpc
export FC=ifort
export CMAKE_C_COMPILER=icc
export CMAKE_CXX_COMPILER=icpc
export CMAKE_Fortran_COMPILER=ifort

# MPI
export MPI_C_COMPILER=mpiicc
export MPI_CXX_COMPILER=mpiicpc
export MPI_Fortran_COMPILER=mpiifort
```

* **The configuration file, sourced to setup the GNU Compilers toolchain**
```
# C/C++/Fortran
export CC=gcc
export CXX=g++
export FC=gfortran
```

* **The linking status of libmkldnn.so after the Intel+MKL+IntelMP compilation**
```
        linux-vdso.so.1 (0x00007ffe3dd73000)
        libiomp5.so => /opt/intel/compilers_and_libraries_2019.0.046/linux/compiler/lib/intel64_lin/libiomp5.so (0x00007f759df90000)
        libmkl_rt.so => /opt/intel/compilers_and_libraries_2019.0.046/linux/mkl/lib/intel64_lin/libmkl_rt.so (0x00007f759d901000)
        libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007f759d57a000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007f759d22e000)
        libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007f759d017000)
        libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f759cdf9000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007f759ca42000)
        libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f759c83e000)
        /usr/lib64/ld-linux-x86-64.so.2 (0x00007f759f213000)
```

* **The linking status of libmkldnn.so after the GNU (no MKL, no IntelMP) compilation**
```
        linux-vdso.so.1 (0x00007ffff6fa7000)
        libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007f2ef5275000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007f2ef4f29000)
        libgomp.so.1 => /usr/lib/libgomp.so.1 (0x00007f2ef4cfb000)
        libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007f2ef4ae4000)
        libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f2ef48c6000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007f2ef450f000)
        /usr/lib64/ld-linux-x86-64.so.2 (0x00007f2ef5dcc000)
        libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f2ef430b000)
```

* **The failed test logfile (in the Intel+MKL+IntelMP case)**
```
21/36 Testing: test_batch_normalization
21/36 Test: test_batch_normalization
Command: "/home/emaballarin/Downloads/mkl-dnn/build/tests/gtests/test_batch_normalization"
Directory: /home/emaballarin/Downloads/mkl-dnn/build/tests/gtests
"test_batch_normalization" start time: Apr 20 03:36 CEST
Output:
----------------------------------------------------------
[==========] Running 140 tests from 8 test cases.
[----------] Global test environment set-up.
[----------] 4 tests from Simple_NC/bnrm_test_float
[ RUN      ] Simple_NC/bnrm_test_float.TestsBnrm/0
OMP: Info #211: KMP_AFFINITY: decoding x2APIC ids.
OMP: Info #209: KMP_AFFINITY: Affinity capable, using global cpuid leaf 11 info
OMP: Info #154: KMP_AFFINITY: Initial OS proc set respected: {0,1,2,3,4,5,6,7}
OMP: Info #156: KMP_AFFINITY: 8 available OS procs
OMP: Info #157: KMP_AFFINITY: Uniform topology
OMP: Info #179: KMP_AFFINITY: 1 packages x 4 cores/pkg x 2 threads/core (4 total cores)
OMP: Info #213: KMP_AFFINITY: OS proc to physical thread map:
OMP: Info #171: KMP_AFFINITY: OS proc 0 maps to package 0 core 0 thread 0 
OMP: Info #171: KMP_AFFINITY: OS proc 4 maps to package 0 core 0 thread 1 
OMP: Info #171: KMP_AFFINITY: OS proc 1 maps to package 0 core 1 thread 0 
OMP: Info #171: KMP_AFFINITY: OS proc 5 maps to package 0 core 1 thread 1 
OMP: Info #171: KMP_AFFINITY: OS proc 2 maps to package 0 core 2 thread 0 
OMP: Info #171: KMP_AFFINITY: OS proc 6 maps to package 0 core 2 thread 1 
OMP: Info #171: KMP_AFFINITY: OS proc 3 maps to package 0 core 3 thread 0 
OMP: Info #171: KMP_AFFINITY: OS proc 7 maps to package 0 core 3 thread 1 
OMP: Info #249: KMP_AFFINITY: pid 18970 tid 18970 thread 0 bound to OS proc set {0}
OMP: Info #249: KMP_AFFINITY: pid 18970 tid 18971 thread 1 bound to OS proc set {1}
OMP: Info #249: KMP_AFFINITY: pid 18970 tid 18972 thread 2 bound to OS proc set {2}
OMP: Info #249: KMP_AFFINITY: pid 18970 tid 18973 thread 3 bound to OS proc set {3}
<end of output>
Test time =   0.23 sec
----------------------------------------------------------
Test Failed.
"test_batch_normalization" end time: Apr 20 03:36 CEST
"test_batch_normalization" time elapsed: 00:00:00
----------------------------------------------------------
```

* **The failed test logfile (in the GNU case)**
```
20/35 Testing: test_batch_normalization
20/35 Test: test_batch_normalization
Command: "/home/emaballarin/Downloads/mkl-dnn-gcc/build/tests/gtests/test_batch_normalization"
Directory: /home/emaballarin/Downloads/mkl-dnn-gcc/build/tests/gtests
"test_batch_normalization" start time: Apr 20 03:54 CEST
Output:
----------------------------------------------------------
[==========] Running 140 tests from 8 test cases.
[----------] Global test environment set-up.
[----------] 4 tests from Simple_NC/bnrm_test_float
[ RUN      ] Simple_NC/bnrm_test_float.TestsBnrm/0
<end of output>
Test time =   0.25 sec
----------------------------------------------------------
Test Failed.
"test_batch_normalization" end time: Apr 20 03:54 CEST
"test_batch_normalization" time elapsed: 00:00:00
----------------------------------------------------------
```

**Comment by emfomenk** [comment]
Hi @emaballarin,

Thanks for the report!
We indeed recently introduced this regression.
Should be fixed very soon.

**Comment by emfomenk** [comment]
Closed with 575b7e860d824a2d3fe901f36f22b00faeaf037f.
